homeassistant/components/ozw/config_flow.py

```diff
@@ -1,24 +1,175 @@
 
+_LOGGER = logging.getLogger(__name__)
+
+CONF_ADDON_DEVICE = "device"
+CONF_ADDON_NETWORK_KEY = "network_key"
+CONF_NETWORK_KEY = "network_key"
+CONF_USB_PATH = "usb_path"
+CONF_USE_ADDON = "use_addon"
 TITLE = "OpenZWave"
 
+ON_SUPERVISOR_SCHEMA = vol.Schema({vol.Optional(CONF_USE_ADDON, default=False): bool})
+
 
 class DomainConfigFlow(config_entries.ConfigFlow, domain=DOMAIN):
     """Handle a config flow for ozw."""
 
     VERSION = 1
     CONNECTION_CLASS = config_entries.CONN_CLASS_LOCAL_PUSH
 
+    def __init__(self):
+        """Set up flow instance."""
+        self.addon_config = None
+        self.addon_info = None
+        self.network_key = None
+        self.usb_path = None
+        # If we install the add-on we should uninstall it on entry remove.
+        self.integration_created_addon = False
+
     async def async_step_user(self, user_input=None):
         """Handle the initial step."""
         if self._async_current_entries():
             return self.async_abort(reason="single_instance_allowed")
+
+        if not self.hass.components.hassio.is_hassio():
+            return self._async_use_mqtt_integration()
+
+        return await self.async_step_on_supervisor()
+
+    @callback
+    def _async_use_mqtt_integration(self):
+        """Handle logic when using the MQTT integration."""
         if "mqtt" not in self.hass.config.components:
             return self.async_abort(reason="mqtt_required")
+
+        return self.async_create_entry(
+            title=TITLE,
+            data={
+                CONF_USB_PATH: self.usb_path,
+                CONF_NETWORK_KEY: self.network_key,
+                CONF_INTEGRATION_CREATED_ADDON: self.integration_created_addon,
+            },
+        )
+
+    async def async_step_on_supervisor(self, user_input=None):
+        """Handle logic when on Supervisor host."""
         if user_input is not None:
-            return self.async_create_entry(title=TITLE, data={})
+            if user_input[CONF_USE_ADDON]:
+                return await self._async_use_addon()
+            return self._async_use_mqtt_integration()
+
+        return self.async_show_form(
+            step_id="on_supervisor", data_schema=ON_SUPERVISOR_SCHEMA
+        )
+
+    async def _async_use_addon(self):
+        """Handle logic when using the OpenZWave add-on."""
+        if await self._async_is_addon_running():
+            return self._async_use_mqtt_integration()
+
+        if await self._async_is_addon_installed():
+            return await self.async_step_start_addon()
+
+        return await self._async_install_addon()
+
+    async def _async_install_addon(self):
+        """Install add-on."""
+        try:
+            await self.hass.components.hassio.async_install_addon("core_zwave")
+        except HassioAPIError as err:
+            _LOGGER.error("Failed to install OpenZWave add-on: %s", err)
+            return self.async_abort(reason="addon_install_failed")
```

Comment:
It feels weird that we're returning an abort result from a function that's not a step. I also don't see this function being used more than once. Should it be broken out or be part of the step ?

Reply:
I'll make it a step. It'll be a step when we start using the show progress helper.